### DISCUSSION TITLE
Falsche Messwerte im Bereich der 400/500 Liter Marken beim Analogzähler (ana-1)

### DISCUSSION THREAD

**Discussion by TopGoalVl** [jomjol/AI-on-the-edge-device#1931] [Q&A] [answered]
Die Messwerte werden, bis auf einige Ausreißer, eigentlich zufriedenstellend erkannt.
 
Massive Probleme gibt es nur im 400 / 500 Liter Bereich vom Analogzähler (ana-1). Das habe ich für mehrere Tage geprüft.
Es sieht so aus, als würde der 500 Liter Bereich viel zu früh aktiv (der Null-Durchlauf betrifft die 400 Liter Marke).  

Der Auszug aus dem Data Log vom 27.01.2023, ab 7:25 Uhr soll das verdeutlichen.

Ich verwende ein 60° Objektiv am ESP32CAM-MB Modul.

Aktuelle Erkennungsmodelle:
dig-1 bis dig-4: dig-class100-0140_s2_q.tflite
ana-1 bis ana-4: ana-cont_1104_s2_q.tflite

Pre Value und negative Werte habe ich in der Config deaktiviert, weil ein zu hoher/falscher Pre Value verhindert, dass sich
die Messergebnisse automatisch regulieren. Das scheint aber nicht zu funktionieren. Oder ich mache etwas falsch oder einen Denkfehler.

Was kann ich machen, um die Erkennung im o. g. Bereich zu verbessen?
Ist die Qualität der Fotos für die Erkennung gut genug?

**Die folgenden Screenshots passen zu den Data Log Einträgen:**

![WzMesswert vom 2023-01-27 07-30-30](https://user-images.githubusercontent.com/119757556/215275826-4482bf9c-6e92-4be3-b429-cc8612004e80.png)
**Null Durchlauf (ana-2) erfolgt, 300 l Marke aktiv.**

![WzMesswert vom 2023-01-27 07-35-21](https://user-images.githubusercontent.com/119757556/215275825-7075f3b0-8303-4d16-82b5-e14b99fae942.png)
**Messwert 419.4048 noch ok, liegt im 400 l Bereich.**

![WzMesswert vom 2023-01-27 07-40-12](https://user-images.githubusercontent.com/119757556/215277465-ce622115-68a6-4d5b-95c7-0c84c49e4676.png)
**Messwert 419.5389 zu hoch, 400 l Bereich noch aktiv!**

![WzMesswert vom 2023-01-27 07-45-04](https://user-images.githubusercontent.com/119757556/215275823-a5f2fb75-1369-4c40-96af-b7b367888711.png)
**Messwert 419.5389 zu hoch, 400 l Bereich noch aktiv!**

![WzMesswert vom 2023-01-27 07-49-55](https://user-images.githubusercontent.com/119757556/215275821-318f6ace-418d-4712-b829-e3f7a53ee73f.png)
**Messwert 419.5389 zu hoch, 400 l Bereich noch aktiv!**

 **                ===> Korrekte Messwerte für die 500 Liter Marke ergaben sich erst ab 16:33 Uhr <===**

![WzScreenshot vom 2023-01-28 15-46-02](https://user-images.githubusercontent.com/119757556/215275819-f9953ff7-8ad5-4516-b2e9-c78338fc466e.png)
**Aktueller Screenshot zur Beurteilung der Erkennungsqualität**

`
Data Log Extrakt
----------------
2023-01-27T07:25:42+0100,main,0419.3343,419.3343,419.3343,0.000124,0.0006,no error,0.0,4.0,1.0,9.0,3.5,3.4,4.3,3.7
2023-01-27T07:30:34+0100,main,0419.3393,419.3393,419.3393,0.001027,0.0050,no error,0.0,4.0,1.0,8.9,3.5,3.8,9.4,3.6
2023-01-27T07:35:24+0100,main,0419.4048,419.4048,419.4048,0.013552,0.0655,no error,0.0,4.0,1.0,9.0,3.9,0.2,4.9,8.8
                                        ===> Null-Durchlauf für 500 Liter noch nicht erfolgt <===
                                              Zu diesem Zeitpunkt ist die 400 Liter Marke aktuell
2023-01-27T07:40:16+0100,main,0419.5327,419.5327,419.5327,0.026281,0.1279,no error,0.0,4.0,1.0,8.9,4.7,3.3,2.6,7.0 
2023-01-27T07:45:07+0100,main,0419.5389,419.5389,419.5389,0.001278,0.0062,no error,0.0,4.0,1.0,8.9,4.7,3.8,9.0,9.0
2023-01-27T07:49:58+0100,main,0419.5389,419.5389,419.5389,0.000000,0.0000,no error,0.0,4.0,1.0,9.0,4.7,3.8,9.0,9.5
2023-01-27T07:54:49+0100,main,0419.5389,419.5389,419.5389,0.000000,0.0000,no error,0.0,4.0,1.0,8.9,4.7,3.8,9.0,9.5
2023-01-27T07:59:40+0100,main,0419.5389,419.5389,419.5389,0.000000,0.0000,no error,0.0,4.0,1.0,9.0,4.7,3.8,9.0,9.5

                ===> Korrekte Messwerte für die 500 Liter Marke ergaben sich erst ab 16:33 Uhr <===

2023-01-27T16:33:58+0100,main,0419.5159,419.5159,419.5159,0.000042,0.0012,no error,0.0,4.0,1.0,9.0,5.4,1.4,5.9,9.0

Aktuelle Config.ini
-------------------
[MakeImage]
LogImageLocation = /log/source
WaitBeforeTakingPicture = 5
LogfileRetentionInDays = 32
Brightness = 0
Contrast = 0
Saturation = 0
LEDIntensity = 45
ImageQuality = 12
ImageSize = VGA
FixedExposure = true

[Alignment]
InitialRotate = 180
InitialMirror = false
SearchFieldX = 20
SearchFieldY = 20
AlignmentAlgo = default
FlipImageSize = false
/config/ref0.jpg 115 110
/config/ref1.jpg 472 105

[Digits]
Model = /config/dig-class100-0140_s2_q.tflite
CNNGoodThreshold = 0.5
LogImageLocation = /log/digit
LogfileRetentionInDays = 32
main.dig1 256 88 36 60 false
main.dig2 308 88 36 60 false
main.dig3 358 88 36 60 false
main.dig4 407 88 36 60 false

[Analog]
Model = /config/ana-cont_1104_s2_q.tflite
LogImageLocation = /log/analog
LogfileRetentionInDays = 32
main.ana1 456 197 102 102 false
main.ana2 396 310 102 102 false
main.ana3 288 353 102 102 false
main.ana4 150 301 102 102 false

[PostProcessing]
main.DecimalShift = 0
main.AnalogDigitalTransitionStart = 9.7
PreValueUse = false
PreValueAgeStartup = 0
AllowNegativeRates = false
main.MaxRateValue = 1
main.MaxRateType = RateChange
main.ExtendedResolution = false
main.IgnoreLeadingNaN = true
ErrorMessage = true
CheckDigitIncreaseConsistency = true

[MQTT]
Uri = mqtt://192.168.188.19:1883
MainTopic = xxxxxxxxxxx/Wasserzähler
ClientID = xxxxxShWz6
user = xxxxxxxx
password = ****************
SetRetainFlag = true
;HomeassistantDiscovery = false
;MeterType = 

;[InfluxDB]
;Uri = undefined
;Database = 
;Measurement = undefined
;user = undefined
;password = undefined

;[GPIO]
;IO0 = input disabled 10 false false 
;IO1 = input disabled 10 false false 
;IO3 = input disabled 10 false false 
;IO4 = built-in-led disabled 10 false false 
;IO12 = input-pullup disabled 10 false false 
;IO13 = input-pullup disabled 10 false false 
LEDType = WS2812
LEDNumbers = 2
LEDColor = 150 150 150

[AutoTimer]
AutoStart = true
Intervall = 4.85

[DataLogging]
DataLogActive = true
DataLogRetentionInDays = 32

[Debug]
Logfile = 4
LogfileRetentionInDays = 16

[System]
TimeZone = CET-1CEST,M3.5.0,M10.5.0/3
TimeServer = de.pool.ntp.org
;AutoAdjustSummertime = false
;Hostname = undefined
SetupMode = false

`

**Comment by TopGoalVl** [comment]
Sorry, der Wert unter dem Screenshot raw_20230127-074012.jpg  muss lauten 419.5327 und ist um 100 Liter zu hoch!

**Comment by andre68723** [comment]
ähnliches Problem bei mir,

587.63 vs 587.73

![ai](https://user-images.githubusercontent.com/16468234/215345370-f7828062-b9a7-4766-af74-3ddbd6c42f1c.png)

**Comment by TopGoalVl** [comment]
Ja, der Sachverhalt ist identisch. Der letzte Null-Durchlauf von ana2 betrifft die 600 Liter Marke bei ana1. Der korrekte Wert ist 587.6352. Eine Differenz von 100 Litern, die nach PreValue übernommen wird und bei mir zu falschen Messwerten führt, die erst nach Stunden wieder ausgeglichen werden. Daher habe ich PreValue im setup deaktiviert. Das funtioniert leider nicht, so wie ich mir das vorstelle, 

Ob NaN wirklich sein muss, oder vermieden werden kann, kann man lange diskutieren. Ich hatte gestern den Fall, dass ein Wechsel von 419m³ auf 420m³ anstand. Führte zu einem Messwertfehler von 60m³ (s. #1944).

Mein System betreibe ich noch in der Test- / Evaluierungsphase. Mal sehen, ob ich am setup optimieren kann. 

Tipp: Für reference1 hast du "m³" gewählt (hatte ich auch so gemacht). Beim nächsten setup nur das "m" wählen und die Rahmen eng anlegen (pixelgenau). Bei "ZR" auch nachbessern. Die werden zu Ausrichtung verwendet und führen vieleicht zu einer besseren Erkennungsrate.

**Comment by caco3** [comment]
nehmt für den analogteil auch ein class100 Model, dann gibt es keine `NaN`.

Wieso der Wechsel zu früh kommt, verstehe ich auch nicht. Ev. versteht @haverland das besser.

**Comment by TopGoalVl** [comment, marked as answer]
Das Setup habe ich wie folgt geändert:
Es werden nur die Parameter aufgeführt, die ich geändert habe. 
(Alter Stand s. config.ini weiter oben)

```
[MakeImage]
LEDIntensity = 48    <==== Helligkeit verbessert, Reflexion hält sich in Grenzen

[Alignment]
SearchFieldX = 30
SearchFieldY = 30
AlignmentAlgo = fast
/config/ref0.jpg 160 115	<==== enger Rahmen, nur das P PI[P]
/config/ref1.jpg 473 112	<==== enger Rahmen, nur das m [m]³

[Digits]
Model = /config/dig-class100-0150_s2_q.tflite
;CNNGoodThreshold = 0.5   <====  (auf Kommentar gestellt, keine Ahnung was dahinter steckt).

[Analog]
Model = /config/ana-class100_0154_s1_q.tflite	<==== Tipp von @caco3 (s. oben). 

[PostProcessing]
main.AnalogDigitalTransitionStart = 9.2	<==== auf default zurück
PreValueUse = true   <==== auf default zurück
PreValueAgeStartup = 10   <==== ausprobieren, default 720 Minuten zu hoch?
main.MaxRateValue = 0.05 <==== prüfen, ob ausreichend
main.MaxRateType = RateChange    <==== prüfen, ob "ChangeAbsoluteRate" nicht besser ist
CheckDigitIncreaseConsistency = false  <==== testen, ob false besser (default=true) 

```
Beim vorherigen Setup habe ich mit: "PreValueUse = false" getestet, weil zu hohe "falsche" PreValues zu Fehlern führen, die sich erst nach Stunden ausgleichen. Besonders extrem, wenn der Fehler abends/nachts auftritt. Das hat aber nichts gebracht. Vielleicht habe ich auch etwas anderes dahinter vermutet, oder es funktioniert nicht richtig.  

Mit dem neuen Setup läuft es wesentlich stabiler. Der anstehende Wechsel von 420 m³ auf 421 m³ wurde ohne Probleme durchgeführt. Alle Null-Durchläufe (ana4 - ana1) wurden richtig erkannt und behandelt. Es gab lediglich zwei negative Ergebnisse (je -100 ml, ana4) in den letzten zwei Tagen.  (Parameter: "AllowNegativeRates = false").   

Mal abwarten, ob es so stabil bleibt.

Anmerkungen oder Korrekturen zum Setup sind willkommen.


![WzMarkerEngerGesetzt 2023-02-01 19-11-33](https://user-images.githubusercontent.com/119757556/216148091-8ea9c850-b5db-4b1d-846a-7628c7586824.png)

![WzSetupGeaendert vom 2023-02-01 11-42-20](https://user-images.githubusercontent.com/119757556/216148101-8a706952-e8c0-475e-81a0-9fa432aab699.png)

**Reply by caco3** [reply]
> Es gab lediglich zwei negative Ergebnisse (je -100 ml, ana4) 

Gemäss meinem Verständnis kann man das umgehen, indem man `ExtendedResolution` auf true setzt, siehe https://github.com/jomjol/AI-on-the-edge-device/pull/1177
